kubelet: fix mixing up runtime classes with runtime handlers

AkihiroSuda opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind bug

#### What this PR does / why we need it:
-  kubelet: fix mixing up runtime classes with runtime handlers
- kubelet: silence "unknown runtime class" errors when unsupported

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #123906
#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
NONE
```

AkihiroSuda committed:
kubelet: fix mixing up runtime classes with runtime handlers

Fix issue 123906

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda committed:
kubelet: silence "unknown runtime class" errors when unsupported

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda commented on pkg/kubelet/nodestatus/setters.go:
```diff
@@ -481,18 +483,41 @@ func GoRuntime() Setter {
+					"runtimeHandlerName", handlerName)
+				continue
+			}
 			node.Status.RuntimeClasses[i] = v1.NodeRuntimeClass{
```
Wondering if we should have rather returned the list of the RuntimeHandlers, if we are still allowed to change the API

@thockin @haircommander @giuseppe

AkihiroSuda replied:
Renamed to RuntimeHandlers

giuseppe reviewed (approved):
LGTM

ffromani commented:
/triage accepted
/priority important-soon

tallclair commented:
/assign

tallclair commented on pkg/kubelet/nodestatus/setters.go:
```diff
@@ -481,20 +483,49 @@ func GoRuntime() Setter {
+		node.Status.RuntimeClasses = []v1.NodeRuntimeClass{}
+		handler, ok := handlersMap[""]
+		if !ok {
+			return fmt.Errorf("failed to look up the default runtime handler")
```
How is this error handled? This case will be hit if it's using a CRI runtime that hasn't implemented the handlers yet.

AkihiroSuda replied:
The code is removed

tallclair commented on pkg/kubelet/nodestatus/setters.go:
```diff
@@ -481,20 +483,49 @@ func GoRuntime() Setter {
+			handlersMap[h.Name] = h
+		}
+
+		classes, err := rcManager.Lister().List(labels.Everything())
```
Does list return a stable order? If not, this list should probably be sorted to ensure that the list in the status is stable.

AkihiroSuda replied:
The code is removed

tallclair commented on pkg/kubelet/nodestatus/setters.go:
```diff
@@ -481,20 +483,49 @@ func GoRuntime() Setter {
+		for _, class := range classes {
+			handler, ok = handlersMap[class.Handler]
+			if !ok {
+				klog.ErrorS(nil, "Failed to look up the runtime handler", "runtimeClassName", class.Name,
```
This is not necessarily an error. A use case is that there is a dedicated node pool for running sandboxed containers, and a certain RuntimeClass/handler is only available in that node pool. In fact, this is a supported use case through the Scheduling mix-in on the RuntimeClass: https://github.com/kubernetes/api/blob/96558b97565e2a9bed0b5cb84db3b09b71d20172/node/v1/types.go#L65

AkihiroSuda replied:
The code is removed

tallclair commented on pkg/kubelet/nodestatus/setters.go:
```diff
@@ -481,20 +483,49 @@ func GoRuntime() Setter {
+		if err != nil {
+			return fmt.Errorf("failed to get the list of the runtime classes: %w", err)
+		}
+		node.Status.RuntimeClasses = []v1.NodeRuntimeClass{}
```
How is this API consumed? Or is it informational only? I wonder if the API should just be a list of RuntimeHandlers instead?

tallclair commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -2445,23 +2445,32 @@ func (kl *Kubelet) cleanupOrphanedPodCgroups(pcm cm.PodContainerManager, cgroupP
 		}
 	}
-	klog.ErrorS(nil, "unknown runtime class", "runtimeClassName", runtimeClassName)
+	klog.ErrorS(nil, "unknown runtime handler", "runtimeHandlerName", runtimeHandlerName)
```
nit
```suggestion
	klog.ErrorS(nil, "Unknown runtime handler", "runtimeHandlerName", runtimeHandlerName)
```

AkihiroSuda replied:
done

ffromani reviewed (commented):
LGTM

tallclair commented on pkg/kubelet/runtimeclass/runtimeclass_manager.go:
```diff
@@ -76,3 +76,8 @@ func (m *Manager) LookupRuntimeHandler(runtimeClassName *string) (string, error)
 }
+
+// Lister returns the runtime class lister.
+func (m *Manager) Lister() nodev1.RuntimeClassLister {
```
Let's just expose a `List()` function that lists everything instead.

AkihiroSuda replied:
The code is removed

tallclair commented on pkg/kubelet/kuberuntime/util/util_test.go:
```diff
@@ -223,7 +223,7 @@ func TestNamespacesForPod(t *testing.T) {
 	} {
 		t.Run(desc, func(t *testing.T) {
-			actual, err := NamespacesForPod(test.input, &kubecontainertest.FakeRuntimeHelper{})
+			actual, err := NamespacesForPod(test.input, &kubecontainertest.FakeRuntimeHelper{}, nil)
```
This should have user namespace coverage.

AkihiroSuda replied:
Yes, but seems beyond the scope of the PR.

cc @giuseppe @rata

rata replied:
I'll open a follow-up PR for that, thanks!

tallclair reviewed (commented):
Thanks for addressing this.

tallclair commented:
/cc @thockin 

For the API discussion.

AkihiroSuda committed:
api: NodeStatus: rename RuntimeClasses to RuntimeHandlers

The runtime classes are apiserver's concept, while the handlers are kubelet's concept.
For NodeStatus, it makes more sense to return the latter ones here.

This commit modifies the following files:

- pkg/apis/core/types.go
- staging/src/k8s.io/api/core/v1/types.go
- pkg/kubelet/nodestatus/setters.go
- pkg/kubelet/kubelet_node_status.go
- pkg/registry/core/node/strategy.go
- test/e2e_node/mount_rro_linux_test.go

Other changes were auto-generated by running `make update`.

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda commented:
/label api-review

liggitt commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -5925,11 +5925,11 @@ type NodeStatus struct {
 	// +optional
 	// +listType=atomic
-	RuntimeClasses []NodeRuntimeClass `json:"runtimeClasses,omitempty" protobuf:"bytes,12,rep,name=runtimeClasses"`
+	RuntimeHandlers []NodeRuntimeHandler `json:"runtimeHandlers,omitempty" protobuf:"bytes,12,rep,name=runtimeHandlers"`
```
this is brand new in 1.30? if so, it's important this field name change land before release

cc @thockin

thockin commented:
API name change LGTM

/approve
/unhold

thockin commented:
Who has the LGTM ?

dims commented:
let's tag @mrunalp !

mrunalp reviewed (approved):
/lgtm

tallclair commented on pkg/registry/core/node/strategy.go:
```diff
@@ -104,7 +104,7 @@ func dropDisabledFields(node *api.Node, oldNode *api.Node) {
 	}
 
 	if !utilfeature.DefaultFeatureGate.Enabled(features.RecursiveReadOnlyMounts) {
```
It seems weird that this is tied to the RecursiveReadOnlyMounts since it's also used by UserNamespaces.

AkihiroSuda replied:
The flag for UserNamespaces is still not included in the Kubernetes core API struct
